solidity/contracts/BTCUtils.sol

```diff
@@ -137,20 +157,29 @@ library BTCUtils {
     /// @param _vin      The vin as a tightly-packed byte array
     /// @param _index    The 0-indexed location of the input to extract
     /// @return          The input as a byte array
-    function extractInputAtIndex(bytes memory _vin, uint8 _index) internal pure returns (bytes memory) {
-        uint256 _len;
+    function extractInputAtIndex(bytes memory _vin, uint256 _index) internal pure returns (bytes memory) {
+        uint256 _varIntDataLen;
+        uint256 _nIns;
+
+        (_varIntDataLen, _nIns) = parseVarInt(_vin);
+        require(_varIntDataLen != ERR_BAD_ARG, "Read overrun during VarInt parsing");
+        require(_index <= _nIns, "Vin read overrun");
+
         bytes memory _remaining;
 
-        uint256 _offset = 1;
+        uint256 _len = 0;
+        uint256 _offset = 1 + _varIntDataLen;
 
-        for (uint8 _i = 0; _i < _index; _i ++) {
+        for (uint256 _i = 0; _i < _index; _i ++) {
             _remaining = _vin.slice(_offset, _vin.length - _offset);
             _len = determineInputLength(_remaining);
+            require(_len != ERR_BAD_ARG, "Bad VarInt in scriptSig");
             _offset = _offset + _len;
         }
 
         _remaining = _vin.slice(_offset, _vin.length - _offset);
         _len = determineInputLength(_remaining);
+        require(_len != ERR_BAD_ARG, "Bad VarInt in scriptSig");
         return _vin.slice(_offset, _len);
     }
```

Comment:
And `_vin.slice(1, _vin.length - 1)` returns what may be a totally valid input

Reply:
Yes, this is the intended functionality. If `index` is 0, it should skip the loop.